File: src2/MicrotonalHelper/operations.js
import { isRequired, isArray } from "./argumentCheck.js";

export function reduce(
  value = isRequired("value"),
  period = isRequired("period")
) {
  if (period == 1) {
    return value;
  }

  if (value >= period) {
    while (value >= period) {
      value = value / period;
    }
  } else if (value > 0 && value < 1) {
    while (value < 1) {
      value = value * period;
    }
  }
  return value;
}

export function ratioToCents(ratio) {
  return Math.log2(ratio) * 1200;
}

export function centsToRatio(cents) {
  var ratio = Math.pow(2, cents / 1200);
  return ratio;
}

export function ratiosDiffInCents(ratio1, ratio2) {
  return Math.abs(ratioToCents(ratio1) - ratioToCents(ratio2));
}

export function getClosestRatioFromScale(targetRatio, scale) {
  isRequired(targetRatio, "targetRatio");
  isRequired(scale, "scale");
  isArray(scale, "scale");

  const diffMap = scale.map((scaleRatio, scaleIndex) => ({
    scaleIndex,
    diffInCents: ratiosDiffInCents(scaleRatio, targetRatio)
  })).sort((a, b) => a.diffInCents - b.diffInCents);

  return diffMap[0];
}

function scaleFindIndexByReducedRatio(scale, ratio) {
  const period = getScalePeriod(scale);
  const reducedRatio = reduceRatio(ratio, period);
  return scale.findIndex(scaleRatio =>
    isEqualRatio(reducedRatio, reduceRatio(scaleRatio, period))
  );
}


export function computeRotations(scaleArray) {
  var result = [];
  if (scaleArray[0] !== 1) {
    scaleArray = [1, ...scaleArray];
  }
  var scaleLength = scaleArray.length - 1;
  var period = scaleArray[scaleArray.length - 1];
  for (let i = 0; i < scaleLength; i++) {
    const tmpRoot = scaleArray[i];
    var rotationArray = [];
    for (let j = 0; j < scaleLength; j++) {
      var idx = i + j + 1;
      var modIdx = idx % scaleLength;
      var idxPow = Math.floor(idx / scaleLength);
      const item = scaleArray[modIdx];
      const fixedItem = Math.pow(period, idxPow) * item;
      rotationArray.push(fixedItem / tmpRoot);
    }
    result.push(rotationArray);
  }
  return result;
}


export function unique(array) {
  var minValue = 0.00000001//0.01;
  var resultArray = [];
  for (let index = 0; index < array.length; index++) {
    const element = array[index];
    var contains = false;

    if (resultArray.length > 0) {
      for (let j = 0; j < resultArray.length; j++) {
        const element2 = resultArray[j];
        if (Math.abs(element2 - element) < minValue) {
          contains = true;
          break;
        }
      }
    }

    if (!contains) {
      resultArray.push(element);
    }
  }
  return resultArray;
};

export function iterator(min, max, increment, callback) {
  if (min >= max) {
      throw Error(
          "The max value must be greater than the min value"
      );
  }
  const iterationsCount = Math.floor((max - min) / increment);
  console.log(`Number of iterations: ${iterationsCount}`);

  var result = [];
  for (let iteration = 0; iteration <= iterationsCount; iteration++) {
      const currentValue = min + increment * iteration;
      const data = callback(currentValue)
      if (data !== undefined)
          result.push(data);
  }
  return result;
}

File: src3/MicrotonalExplorer/MicrotonalHelpers/Operations.cs
using System.Collections.Immutable;
using System.Linq;

namespace MicrotonalExplorer;

public static class Operations
{

    public static float GetMinInterval(float[] scale)
    {
        var minInterval = scale[scale.Length - 1];
        for (var index = 1; index < scale.Length; index++)
        {
            var interval = scale[index] / scale[index - 1];
            if (interval < minInterval)
            {
                minInterval = interval;
            }
        }
        return minInterval;
    }

    public static float[][] ComputeRotations(float[] scaleArray)
    {
        // if (scaleArray[0] != 1)
        // {
        //     scaleArray = [1, .. scaleArray];
        // }
        var result = new float[scaleArray.Length][];
        var scaleLength = scaleArray.Length;
        var period = scaleArray[scaleArray.Length - 1];
        float tmpRoot = 1;
        for (var i = 0; i < scaleLength; i++)
        {
            var rotationArray = new float[scaleArray.Length];
            for (var j = 0; j < scaleLength; j++)
            {
                int idx = i + j;
                var modIdx = idx % scaleLength;
                var idxPow = (int)Math.Floor((float)idx / scaleLength);
                var item = scaleArray[modIdx];
                var fixedItem = (float)Math.Pow(period, idxPow) * item;
                rotationArray[j] = fixedItem / tmpRoot;
            }
            result[i] = rotationArray;            
            tmpRoot = scaleArray[i];
        }
        return result;
    }


    public static float RatioToCents(float ratio)
    {
        return (float)Math.Log2(ratio) * 1200;
    }

    public static float CentsToRatio(float cents)
    {
        var ratio = (float)Math.Pow(2, cents / 1200);
        return ratio;
    }

    public static float RatiosDiffInCents(float ratio1, float ratio2)
    {
        return Math.Abs(RatioToCents(ratio1) - RatioToCents(ratio2));
    }

    /// <summary>
    /// Reduces a value to be within the range [1, period) by dividing or multiplying by the period.
    /// If period is 1, returns the value unchanged.
    /// If value >= period, divides by period until value < period.
    /// If 0 < value < 1, multiplies by period until value >= 1.
    /// </summary>
    /// <param name="value">The value to reduce</param>
    /// <param name="period">The period to use for reduction</param>
    /// <returns>The reduced value within [1, period)</returns>
    public static float Reduce(float value, float period)
    {
        if (period == 1)
        {
            return value;
        }

        if (value >= period)
        {
            while (value >= period)
            {
                value = value / period;
            }
        }
        else if (value > 0 && value < 1)
        {
            while (value < 1)
            {
                value = value * period;
            }
        }
        
        return value;
    }

    public static ClosestRationFromScaleResult GetClosestRatioFromScale(float targetRatio, float[] scale)
    {
        float minDiff = 1200;
        var scaleIndex = 0;
        for (int i = 0; i < scale.Length; i++)
        {
            var scaleRatio = scale[i];
            var diff = RatiosDiffInCents(scaleRatio, targetRatio);
            if(diff < minDiff){
                minDiff = diff;
                scaleIndex = i;
            }
        }

        return new ClosestRationFromScaleResult
        (
            scaleIndex,
            minDiff,
            targetRatio,
            scale[scaleIndex]
        );
    }

/// <summary>
/// Verify if all target ratios are present inside the scale considering the tolerance. It returns an empty list if at least one target is not matched.
/// </summary>
/// <param name="targetRatios"></param>
/// <param name="scale"></param>
/// <param name="toleranceInCents"></param>
/// <returns></returns>
    public static List<ClosestRationFromScaleResult> FullMatchTargetRatiosToScale(float[] targetRatios, float[] scale, float toleranceInCents)
    {
        var closestData = new List<ClosestRationFromScaleResult>();
        int closestCount = 0;
        for (int targetIndex = 0; targetIndex < targetRatios.Length; targetIndex++)
        {
            var targetRatio = targetRatios[targetIndex];
            var targetResult = GetClosestRatioFromScale(
               targetRatio,
               scale
             );
            if (targetResult.DiffInCents <= toleranceInCents)
            {
                closestCount++;
                closestData.Add(targetResult);
            }
            else{
                return new List<ClosestRationFromScaleResult>();
            }
        }

        return closestData;
    }
}

public record ClosestRationFromScaleResult(int ScaleIndex, float DiffInCents, float TargetScaleRatio, float ClosestScaleRatio);
